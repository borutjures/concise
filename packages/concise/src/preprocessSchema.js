// @flow

import { clone, merge, set as timmSet, setIn } from 'timm';
import type { Schema } from 'concise-types';

const preprocess = (schema: Schema): Schema => {
  const models = {};

  // Process includes
  Object.keys(schema.models).forEach(modelName => {
    const model0 = schema.models[modelName];
    if (model0.includeOnly) return;
    let model = clone(model0);
    if (model.includes) {
      Object.keys(model.includes).forEach(includeName => {
        const include = schema.models[includeName];
        if (!include) throw new Error(`INCLUDE_NOT_FOUND ${modelName}/${includeName}`);
        const { fields, relations } = include;
        model = merge(model, {
          fields: fields ? merge(fields, model.fields) : undefined,
          relations: relations ? merge(relations, model.relations) : undefined,
        });
      });
    }
    delete model.includes;
    models[modelName] = model;
  });

  // Process relation field types
  Object.keys(models).forEach(modelName => {
    let { relations } = models[modelName];
    Object.keys(relations || {}).forEach(relationName => {
      if (relations[relationName] === true) {
        relations = timmSet(relations, relationName, { model: relationName });
      }
      if (relations[relationName].model == null) {
        relations = setIn(relations, [relationName, 'model'], relationName);
      }
      const relation = relations[relationName];
      const relatedModel = models[relation.model];
      if (!relatedModel) {
        throw new Error(`RELATED_MODEL_NOT_FOUND ${modelName}/${relationName}/${relation.model}`);
      }
      const idField = relatedModel && relatedModel.fields && relatedModel.fields.id;
      if (!idField) {
        throw new Error(`ID_FIELD_NOT_FOUND ${modelName}/${relationName}/${relation.model}`);
      }
      relations = setIn(relations, [relationName, 'type'], idField.type);
    });
    models[modelName] = timmSet(models[modelName], 'relations', relations);
  });

  return { models };
};

export default preprocess;
