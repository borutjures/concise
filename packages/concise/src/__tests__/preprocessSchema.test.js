/* eslint-env jest */

import preprocessSchema from '../preprocessSchema';

describe('preprocessSchema', () => {
  it('should respect unknown schema attributes', () => {
    const processedSchema = preprocessSchema({
      models: {
        person: {
          fields: {
            name: { type: 'string', foo: 3 },
          },
          foo: 3,
        },
      },
      schemaFoo: 5,
    });
    expect(processedSchema).toMatchSnapshot();
  });

  it('should process included fields', () => {
    const processedSchema = preprocessSchema({
      models: {
        common: {
          includeOnly: true,
          fields: {
            id: { type: 'string' },
          },
        },
        person: {
          includes: { common: true },
          fields: {
            name: { type: 'string' },
          },
        },
      },
    });
    expect(processedSchema).toMatchSnapshot();
  });

  it('should complete a shorthand relation', () => {
    const processedSchema = preprocessSchema({
      models: {
        person: {
          fields: {
            id: { type: 'string' },
            name: { type: 'string' },
          },
        },
        post: {
          fields: {
            id: { type: 'string' },
            title: { type: 'string' },
          },
          relations: {
            person: true,
          },
        },
      },
    });
    expect(processedSchema).toMatchSnapshot();
  });
});
