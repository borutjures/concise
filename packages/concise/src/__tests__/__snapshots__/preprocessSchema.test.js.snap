// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preprocessSchema allows custom inverse relations (name) 1`] = `
Object {
  "models": Object {
    "person": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
      },
      "relations": Object {
        "createdPosts": Object {
          "fkName": "createdPostIds",
          "inverseName": "person",
          "isInverse": true,
          "isPlural": true,
          "model": "post",
          "type": "string",
          "validations": Object {},
        },
      },
    },
    "post": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
      },
      "relations": Object {
        "person": Object {
          "fkName": "personId",
          "inverseName": "createdPosts",
          "isInverse": false,
          "isPlural": false,
          "model": "person",
          "type": "string",
          "validations": Object {},
        },
      },
    },
  },
}
`;

exports[`preprocessSchema allows custom inverse relations (singular) 1`] = `
Object {
  "models": Object {
    "address": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
      },
      "relations": Object {
        "person": Object {
          "fkName": "personId",
          "inverseName": "address",
          "isInverse": false,
          "isPlural": false,
          "model": "person",
          "type": "string",
          "validations": Object {},
        },
      },
    },
    "person": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
      },
      "relations": Object {
        "address": Object {
          "fkName": "addressId",
          "inverseName": "person",
          "isInverse": true,
          "isPlural": false,
          "model": "address",
          "type": "string",
          "validations": Object {},
        },
      },
    },
  },
}
`;

exports[`preprocessSchema allows disabling inverse relations 1`] = `
Object {
  "models": Object {
    "person": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
      },
      "relations": Object {},
    },
    "post": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
      },
      "relations": Object {
        "person": Object {
          "fkName": "personId",
          "inverseName": null,
          "isInverse": false,
          "isPlural": false,
          "model": "person",
          "type": "string",
          "validations": Object {},
        },
      },
    },
  },
}
`;

exports[`preprocessSchema completes custom relations 1`] = `
Object {
  "models": Object {
    "person": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
      },
      "relations": Object {
        "posts": Object {
          "fkName": "postIds",
          "inverseName": "author",
          "isInverse": true,
          "isPlural": true,
          "model": "post",
          "type": "string",
          "validations": Object {},
        },
      },
    },
    "post": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
      },
      "relations": Object {
        "author": Object {
          "fkName": "authorId",
          "inverseName": "posts",
          "isInverse": false,
          "isPlural": false,
          "model": "person",
          "type": "string",
          "validations": Object {
            "required": true,
          },
        },
      },
    },
  },
}
`;

exports[`preprocessSchema processes a shorthand relation 1`] = `
Object {
  "models": Object {
    "person": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
      },
      "relations": Object {
        "posts": Object {
          "fkName": "postIds",
          "inverseName": "person",
          "isInverse": true,
          "isPlural": true,
          "model": "post",
          "type": "string",
          "validations": Object {},
        },
      },
    },
    "post": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
      },
      "relations": Object {
        "person": Object {
          "fkName": "personId",
          "inverseName": "posts",
          "isInverse": false,
          "isPlural": false,
          "model": "person",
          "type": "string",
          "validations": Object {},
        },
      },
    },
  },
}
`;

exports[`preprocessSchema processes included fields 1`] = `
Object {
  "models": Object {
    "person": Object {
      "fields": Object {
        "id": Object {
          "type": "string",
        },
        "name": Object {
          "type": "string",
        },
      },
      "relations": Object {},
    },
  },
}
`;

exports[`preprocessSchema respects unknown schema attributes 1`] = `
Object {
  "models": Object {
    "person": Object {
      "fields": Object {
        "name": Object {
          "foo": 3,
          "type": "string",
        },
      },
      "foo": 3,
      "relations": Object {},
    },
  },
  "schemaFoo": 5,
}
`;
